File: server/import-clients.js
import * as XLSX from 'xlsx';
import { db } from './db.js';

export const TEMPLATE_HEADERS = [
  'ID_Salesforce',
  'Nombre',
  'Telefono',
  'Email',
  'Direccion',
  'Localidad',
  'Provincia',
  'Notas',
];

const ALIASES = {
  ID_Salesforce: ['id_salesforce', 'id', 'salesforce id', 'account id', 'accountid', 'case number', 'casenumber', 'id externo', 'external id'],
  Nombre: ['nombre', 'name', 'account name', 'accountname', 'razon social', 'cliente'],
  Telefono: ['telefono', 'phone', 'mobile', 'mobilephone', 'celular'],
  Email: ['email', 'correo'],
  Direccion: ['direccion', 'dirección', 'street', 'billingstreet', 'address', 'domicilio'],
  Localidad: ['localidad', 'city', 'billingcity', 'ciudad'],
  Provincia: ['provincia', 'state', 'billingstate'],
  Notas: ['notas', 'notes', 'description', 'descripcion', 'descripción'],
};

function norm(value) {
  return String(value || '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .trim()
    .toLowerCase();
}

function mapRow(row) {
  const keys = Object.keys(row);
  const pick = (field) => {
    const aliases = ALIASES[field];
    const found = keys.find((k) => aliases.includes(norm(k)));
    const val = found ? row[found] : '';
    return val == null ? '' : String(val).trim();
  };
  return {
    external_id: pick('ID_Salesforce'),
    name: pick('Nombre'),
    phone: pick('Telefono'),
    email: pick('Email'),
    address: pick('Direccion'),
    locality: pick('Localidad'),
    province: pick('Provincia'),
    notes: pick('Notas'),
  };
}

function localityText(name, province) {
  if (!name) return null;
  if (province) return `${name}, ${province}`;
  return name;
}

export function buildTemplateBuffer() {
  const sample = [{
    ID_Salesforce: '001XX000003ABC',
    Nombre: 'Juan Pérez',
    Telefono: '11 5555-0000',
    Email: 'juan@correo.com',
    Direccion: 'Av. Siempre Viva 742',
    Localidad: 'Quilmes',
    Provincia: 'Buenos Aires',
    Notas: 'Exportado desde Salesforce',
  }];
  const sheet = XLSX.utils.json_to_sheet(sample, { header: TEMPLATE_HEADERS });
  const wb = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(wb, sheet, 'Clientes');
  return XLSX.write(wb, { type: 'buffer', bookType: 'xlsx' });
}

export function importClientsFromBuffer(buffer, providerId) {
  const wb = XLSX.read(buffer, { type: 'buffer' });
  const sheet = wb.Sheets[wb.SheetNames[0]];
  const rows = XLSX.utils.sheet_to_json(sheet, { defval: '' });
  const find = db.prepare(
    'SELECT id FROM clients WHERE provider_id = ? AND external_id = ? LIMIT 1'
  );
  const insert = db.prepare(
    `INSERT INTO clients (provider_id, external_id, name, phone, email, locality, address, notes, source)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'excel')`
  );
  const update = db.prepare(
    `UPDATE clients SET name=?, phone=?, email=?, locality=?, address=?, notes=?, source='excel'
     WHERE id=?`
  );

  let created = 0;
  let updated = 0;
  const errors = [];

  for (const [i, raw] of rows.entries()) {
    const row = mapRow(raw);
    if (!row.name) {
      errors.push(`Fila ${i + 2}: falta Nombre`);
      continue;
    }
    const locality = localityText(row.locality, row.province);
    const ext = row.external_id || null;
    const existing = ext ? find.get(providerId, ext) : null;
    if (existing) {
      update.run(row.name, row.phone || null, row.email || null, locality, row.address || null, row.notes || null, existing.id);
      updated += 1;
    } else {
      insert.run(providerId, ext, row.name, row.phone || null, row.email || null, locality, row.address || null, row.notes || null);
      created += 1;
    }
  }

  return { created, updated, errors, total: rows.length };
}

File: backend/src/services/import_clients.go
package services

import (
	"bytes"
	"context"
	"database/sql"
	"errors"
	"strconv"
	"strings"
	"unicode"

	"taller-gestion/backend/src/models"
	"taller-gestion/backend/src/utils"

	"github.com/xuri/excelize/v2"
)

var templateHeaders = []string{
	"ID_Salesforce", "Nombre", "Telefono", "Email", "Direccion", "Localidad", "Provincia", "Notas",
}

var fieldAliases = map[string][]string{
	"ID_Salesforce": {"id_salesforce", "id", "salesforce id", "account id", "accountid", "case number", "casenumber", "id externo", "external id"},
	"Nombre":        {"nombre", "name", "account name", "accountname", "razon social", "cliente"},
	"Telefono":      {"telefono", "phone", "mobile", "mobilephone", "celular"},
	"Email":         {"email", "correo"},
	"Direccion":     {"direccion", "street", "billingstreet", "address", "domicilio"},
	"Localidad":     {"localidad", "city", "billingcity", "ciudad"},
	"Provincia":     {"provincia", "state", "billingstate"},
	"Notas":         {"notas", "notes", "description", "descripcion"},
}

func normHeader(value string) string {
	var b strings.Builder
	for _, r := range strings.ToLower(strings.TrimSpace(value)) {
		switch r {
		case 'á', 'à', 'ä', 'â':
			b.WriteRune('a')
		case 'é', 'è', 'ë', 'ê':
			b.WriteRune('e')
		case 'í', 'ì', 'ï', 'î':
			b.WriteRune('i')
		case 'ó', 'ò', 'ö', 'ô':
			b.WriteRune('o')
		case 'ú', 'ù', 'ü', 'û':
			b.WriteRune('u')
		case 'ñ':
			b.WriteRune('n')
		default:
			if unicode.IsLetter(r) || unicode.IsDigit(r) || r == ' ' {
				b.WriteRune(r)
			}
		}
	}
	return b.String()
}

func (s *Services) BuildClientsTemplate() ([]byte, error) {
	f := excelize.NewFile()
	sheet := f.GetSheetName(0)
	_ = f.SetSheetName(sheet, "Clientes")
	sheet = "Clientes"
	sample := []string{
		"001XX000003ABC", "Juan Pérez", "11 5555-0000", "juan@correo.com",
		"Av. Siempre Viva 742", "Quilmes", "Buenos Aires", "Exportado desde Salesforce",
	}
	for i, h := range templateHeaders {
		cell, _ := excelize.CoordinatesToCellName(i+1, 1)
		_ = f.SetCellValue(sheet, cell, h)
		cell2, _ := excelize.CoordinatesToCellName(i+1, 2)
		_ = f.SetCellValue(sheet, cell2, sample[i])
	}
	var buf bytes.Buffer
	if err := f.Write(&buf); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

type ImportResult struct {
	Created int      `json:"created"`
	Updated int      `json:"updated"`
	Errors  []string `json:"errors"`
	Total   int      `json:"total"`
}

type importRow struct {
	ExternalID string
	Name       string
	Phone      string
	Email      string
	Address    string
	Locality   string
	Province   string
	Notes      string
}

func mapImportRow(headers []string, values []string) importRow {
	lookup := map[string]string{}
	for i, h := range headers {
		val := ""
		if i < len(values) {
			val = strings.TrimSpace(values[i])
		}
		lookup[normHeader(h)] = val
	}
	pick := func(field string) string {
		for _, alias := range fieldAliases[field] {
			if v, ok := lookup[alias]; ok {
				return v
			}
		}
		if v, ok := lookup[normHeader(field)]; ok {
			return v
		}
		return ""
	}
	return importRow{
		ExternalID: pick("ID_Salesforce"),
		Name:       pick("Nombre"),
		Phone:      pick("Telefono"),
		Email:      pick("Email"),
		Address:    pick("Direccion"),
		Locality:   pick("Localidad"),
		Province:   pick("Provincia"),
		Notes:      pick("Notas"),
	}
}

func localityText(name, province string) *string {
	if name == "" {
		return nil
	}
	if province != "" {
		s := name + ", " + province
		return &s
	}
	return &name
}

func (s *Services) ClientsTemplate(ctx context.Context, claims *utils.Claims) ([]byte, error) {
	if !utils.HasRole(claims.Role, "admin", "coordinador") {
		return nil, ErrForbidden
	}
	return s.BuildClientsTemplate()
}

func (s *Services) ImportClients(ctx context.Context, claims *utils.Claims, providerID int64, data []byte) (*ImportResult, error) {
	if !utils.HasRole(claims.Role, "admin", "coordinador") {
		return nil, ErrForbidden
	}
	if providerID <= 0 {
		return nil, BadRequest("Elegí el prestador/proveedor de origen")
	}
	f, err := excelize.OpenReader(bytes.NewReader(data))
	if err != nil {
		return nil, BadRequest("No se pudo leer el Excel. Usá la plantilla del sistema.")
	}
	defer f.Close()
	sheets := f.GetSheetList()
	if len(sheets) == 0 {
		return nil, BadRequest("No se pudo leer el Excel. Usá la plantilla del sistema.")
	}
	rows, err := f.GetRows(sheets[0])
	if err != nil || len(rows) == 0 {
		return nil, BadRequest("No se pudo leer el Excel. Usá la plantilla del sistema.")
	}
	headers := rows[0]
	result := &ImportResult{Errors: []string{}, Total: len(rows) - 1}

	for i, raw := range rows[1:] {
		row := mapImportRow(headers, raw)
		if row.Name == "" {
			result.Errors = append(result.Errors, "Fila "+strconv.Itoa(i+2)+": falta Nombre")
			continue
		}
		loc := localityText(row.Locality, row.Province)
		phone := utils.OptStr(row.Phone)
		email := utils.OptStr(row.Email)
		address := utils.OptStr(row.Address)
		notes := utils.OptStr(row.Notes)
		ext := utils.OptStr(row.ExternalID)

		var existing *models.Client
		if ext != nil {
			existing, err = s.Repos.FindClientByProviderExternal(ctx, providerID, *ext)
			if err != nil {
				if !errors.Is(err, sql.ErrNoRows) {
					return nil, err
				}
				existing = nil
			}
		}
		if existing != nil {
			if err := s.Repos.UpdateClientImport(ctx, existing.ID, row.Name, phone, email, loc, address, notes); err != nil {
				return nil, err
			}
			result.Updated++
		} else {
			pid := providerID
			client := &models.Client{
				ProviderID: &pid,
				ExternalID: ext,
				Name:       row.Name,
				Phone:      phone,
				Email:      email,
				Locality:   loc,
				Address:    address,
				Notes:      notes,
				Source:     "excel",
			}
			if err := s.Repos.InsertClient(ctx, client); err != nil {
				return nil, err
			}
			result.Created++
		}
	}
	return result, nil
}
